### PR TITLE
[INLONG-600]Multiplexed connection pool for Go sdk

### PR DESCRIPTION
**Pull request by TszKitLo40**

### CODE REVIEW

**Comment by TszKitLo40** [comment]
@gosonzhang @charlely PTAL

**Review by gosonzhang — APPROVED** [review]
+1

File: tubemq-client-twins/tubemq-client-go/pool/multiplexed.go [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pool
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/pool/multiplexed.go:18]: Please comment the package.

Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/pool/multiplexed.go:18]: Can use multiplexedpool to this package. pool maybe buffer or io pool ?

File: tubemq-client-twins/tubemq-client-go/pool/multiplexed.go [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+)
+
+const (
+	Initial int = iota
+	Connected
+	Closing
+	Closed
+)
+
+var queueSize = 10000
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/pool/multiplexed.go:53]: Can use const.

File: tubemq-client-twins/tubemq-client-go/pool/multiplexed.go [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+func New() *Multiplexed {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/pool/multiplexed.go:55]: Multiplexed need name Pool, if this package name multiplexedpool. if use is multiplexedpool.Pool.

Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/pool/multiplexed.go:55]: Suggested code order:
1.New pool.
2.pool, and pool.Get.
3.MultiplexedConnection.
4.Connection.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:17]: Package comment.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	serialNoLen           uint32 = 4
+	beginTokenLen         uint32 = 4
+)
+
+type TransportResponse interface {
+	GetSerialNo() uint32
+	GetResponseBuf() []byte
+}
+
+type Decoder interface {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:44]: Exportable variables and functions require comments.

File: tubemq-client-twins/tubemq-client-go/pool/multiplexed.go (hunk 3 above)
Comment (TszKitLo40) [reply tubemq-client-twins/tubemq-client-go/pool/multiplexed.go]: I have changed the file and package name to multiplexing.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"io"
+)
+
+const (
+	RPCProtocolBeginToken uint32 = 0xFF7FF4FE
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:31]: Exportable variables and functions require comments.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	maxBufferSize         int    = 128 * 1024
+	defaultMsgSize        int    = 4096
+	dataLen               uint32 = 4
+	listSizeLen           uint32 = 4
+	serialNoLen           uint32 = 4
+	beginTokenLen         uint32 = 4
+)
+
+// TransportResponse is the abstraction of the transport response.
+type TransportResponse interface {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:43]: You can use Response class name, becase if use like codec.Response.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// Decoder is the abstraction of the decoder which is used to decode the response.
+type Decoder interface {
+	// Decode will decode the response to frame head and body.
+	Decode() (TransportResponse, error)
+}
+
+// TubeMQDecoder is the implementation of the decoder of response from TubeMQ.
+type TubeMQDecoder struct {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:57]: TubeMQDecoder need other file to writer.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	serialNoLen           uint32 = 4
+	beginTokenLen         uint32 = 4
+)
+
+// TransportResponse is the abstraction of the transport response.
+type TransportResponse interface {
+	// GetSerialNo returns the `serialNo` of the corresponding request.
+	GetSerialNo() uint32
+	// GetResponseBuf returns the body of the response.
+	GetResponseBuf() []byte
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:47]: You can rename GetBuffer, becase this interface is Response.

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+func (t *TubeMQDecoder) Decode() (TransportResponse, error) {
+	num, err := io.ReadFull(t.reader, t.msg[:frameHeadLen])
+	if err != nil {
+		return nil, err
+	}
+	if num != int(frameHeadLen) {
+		return nil, errors.New("framer: read frame header num invalid")
+	}
+	token := binary.BigEndian.Uint32(t.msg[:beginTokenLen])
+	if token != RPCProtocolBeginToken {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:82]: Can merge if binary.BigEndian.Uint32(t.msg[:beginTokenLen]) != RPCProtocolBeginToken {}

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+		}
+
+		s := int(binary.BigEndian.Uint32(size))
+		if totalLen+s > len(t.msg) {
+			data := t.msg[:totalLen]
+			t.msg = make([]byte, totalLen+s)
+			copy(t.msg, data[:])
+		}
+
+		num, err = io.ReadFull(t.reader, t.msg[totalLen:totalLen+s])
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:105]: Can use if num, err = io.ReadFull(t.reader, t.msg[totalLen:totalLen+s]); err != nil {
   return err
}

File: tubemq-client-twins/tubemq-client-go/codec/codec.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+			return nil, err
+		}
+		if n != int(dataLen) {
+			return nil, errors.New("framer: read invalid size")
+		}
+
+		s := int(binary.BigEndian.Uint32(size))
+		if totalLen+s > len(t.msg) {
+			data := t.msg[:totalLen]
+			t.msg = make([]byte, totalLen+s)
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/codec.go:101]: Perhaps more space is needed in the future, and a multiple growth plan should be used.

File: tubemq-client-twins/tubemq-client-go/multiplexing/multiplexing.go [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+		}
+		serialNo := rsp.GetSerialNo()
+		c.mu.RLock()
+		mc, ok := c.connections[serialNo]
+		c.mu.RUnlock()
+		if !ok {
+			continue
+		}
+		mc.reader.recv <- rsp
+		mc.conn.remove(rsp.GetSerialNo())
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/multiplexing/multiplexing.go:350]: Can use like this, mc.conn.remove(serialNo).

File: tubemq-client-twins/tubemq-client-go/codec/tubemq_codec.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+func (t TubeMQResponse) GetSerialNo() uint32 {
+	return t.serialNo
+}
+
+// GetResponseBuf will return the body of Response.
+func (t TubeMQResponse) GetBuffer() []byte {
+	return t.Buffer
+}
+
+func max(x, y int) int {
```
Comment (charlely) [inline tubemq-client-twins/tubemq-client-go/codec/tubemq_codec.go:133]: You can use the math library directly

**Review by charlely — COMMENTED** [review]
LGTM

File: tubemq-client-twins/tubemq-client-go/codec/tubemq_codec.go (hunk 14 above)
Comment (TszKitLo40) [reply tubemq-client-twins/tubemq-client-go/codec/tubemq_codec.go]: Fixed

**Review by charlely — APPROVED** [review]
LGTM

**Pull request merged**